processing/src/main/java/org/apache/druid/segment/column/LexicographicalRangeIndex.java

```diff
@@ -39,18 +39,18 @@ default BitmapColumnIndex forRange(
       boolean endStrict
   )
   {
-    return forRange(startValue, startStrict, endValue, endStrict, (index) -> true);
+    return forRange(startValue, startStrict, endValue, endStrict, null);
   }
 
   /**
    * Get an {@link BitmapColumnIndex} corresponding to the values supplied in the specified range whose dictionary ids
-   * also match some predicate, such as to match a prefix
+   * also match some predicate, such as to match a prefix. Null matchers are equivalent to always-true.
    */
   BitmapColumnIndex forRange(
       @Nullable String startValue,
       boolean startStrict,
       @Nullable String endValue,
       boolean endStrict,
-      Predicate<String> matcher
+      @Nullable Predicate<String> matcher
```

Comment:
What do you think about instead of making this `@Nullable` to remove the `default` implementation of the version of `forRange` which does not take a predicate matcher? I think that is actually the code path your benchmarks were measuring due to how you changed `LikeFilter`, since `DictionaryEncodedStringIndexSupplier` actually does implement that other version.

The reason it had a default implementation is a left-over artifact of #12315 where all of the methods had to be implemented on `BitmapIndex`, meaning a lot of silly implementations were required for things that didn't support those kinds of indexes, so the `default` implementation was a laziness on my part.

I think the implementation of the version that doesn't have the matcher should be slightly better than the `@Nullable` matcher since it can avoid the checks. Though its probably small fries when computing indexes, everything counts.

I think for `ListFilteredVirtualColumn` it would be ok to just use the old 'default' implementation for its version, since its the only other thing that implements this index interface.

Alternatively, we could remove the version that doesn't take a matcher argument and make this version with `@Nullable` the only method, though I think that's slightly less cool because having it split (and not having a default) sort of pushes implementors towards providing the more optimal predicate-less implementation. If we do go this route, we should change `DictionaryEncodedStringIndexSupplier` to make the version without ranges private and just use that when the matcher is null.

This confusion was my fault I think, this PR made it more obvious that the current state isn't the most intuitive on what to call or implement.

Reply:
> What do you think about instead of making this @Nullable to remove the default implementation of the version of forRange which does not take a predicate matcher?

☝️  I did this one